feat: Update SMT Circuit class and add gate relaxation functionality

Sarkoxed opened the pull request:
This pr adds gate relaxation functionality to Circuit class. It substitutes subcircuits with their "arithmetic" equivalents.
Also it adds `util` directory.

### Util

- `default_model` and `default_model_single` will output pretty formatted result of Solver work
- `smt_timer` will measure the amount of time that it took to solve something
- `base4` provides you with accumulators and base4 decomposition of a number

### Terms

- Added `isBitVector()` method that foreshadows future `BitVector` api

### Subcircuits

- functions in `subcircuits.cpp` now return new structure: `CircuitProps`. 
- `CircuitProps` consists of 
    -  `start_gate` indicates where to start subcircuit comparison
    - `num_gates` tells a pure subcircuit size
    - `idxs` and and `gate_idxs` indicate where to search for the values that are used in optimized operation

Also added tests to confirm that all the indices 

### Circuit

- Added members `bool optimizations` and `map<uint32_t, bool> optimized` to keep track of optimized out variables that should not be considered by solver
- Added functions `handle_range_constraint`, `handle_logic_constraint`. They perform binary search to indicate the bit_length of specific operation and apply optimization to subcircuit
- Added cache for subcircuits to avoid recomputing them each time

Tests check that

- all the `assert_equal`'d variables handled properly during circuit initialization
- `range_constraints` are optimized properly

Sarkoxed committed:
Constraint system relaxing

Sarkoxed committed:
Functions to help during testing

Sarkoxed committed:
Test new subcircuits and find offsets

Sarkoxed committed:
A little refactoring

Sarkoxed committed:
isBitVector static method

Sarkoxed committed:
Major changes in subcircuit

Now it returns a structure that contains valuable information for
further optimizations

- start_gate indicates where to start subcircuit comparison
- num_gates tells a pure subcircuit size
- idxs and and gate_idxs indicates where to search for the values that
  are used in optimized operation

Sarkoxed committed:
Finally circuit optimizations

Sarkoxed committed:
Gate relaxation, but using binary search instead

Sarkoxed committed:
Falses

Sarkoxed committed:
minor changes

Sarkoxed committed:
Typo

Sarkoxed committed:
Yet another typo

Rumata888 committed:
Merge branch 'master' into as/smt-update-circuit-2

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit.hpp:
```diff
@@ -168,6 +186,168 @@ void Circuit<FF>::univariate_handler(bb::fr q_m, bb::fr q_1, bb::fr q_2, bb::fr
+    while (beg <= end) {
+        p = (end + beg) / 2;
+
+        xor_props = get_standard_logic_circuit(p * 2, true);
```
Could you please add comments here? The fact that you are constructing circuits and then comparing the value is not intuitive (also you are constructing the circuits every time. It would probably be better to cache them after 1 time)

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit.hpp:
```diff
@@ -188,19 +380,23 @@ template <typename FF> size_t Circuit<FF>::prepare_gates(size_t cursor)
+    optimized[w_r] = false;
+    optimized[w_o] = false;
+
+    bool univariate_flag = true;
```
This definitely needs an explanation

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit.hpp:
```diff
@@ -168,6 +186,168 @@ void Circuit<FF>::univariate_handler(bb::fr q_m, bb::fr q_1, bb::fr q_2, bb::fr
 
+/**
+ * @brief Relaxes univariate logic constraints(AND/XOR).
+ *
```
This function definitely needs a description

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit.hpp:
```diff
@@ -168,6 +186,168 @@ void Circuit<FF>::univariate_handler(bb::fr q_m, bb::fr q_1, bb::fr q_2, bb::fr
+    return res;
+}
+
+template <typename FF> size_t Circuit<FF>::range_constraint_handler(size_t cursor)
```
Please add description. Also the name "handler" would be logical with a class. With a function you want it to be a verb form (mb handle_range_constraints or transform_range_constraints)

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit.hpp:
```diff
@@ -168,6 +186,168 @@ void Circuit<FF>::univariate_handler(bb::fr q_m, bb::fr q_1, bb::fr q_2, bb::fr
+    auto res = static_cast<size_t>(-1);
+
+    CircuitProps range_props;
+    for (size_t odd = 0; odd < 2; odd++) {
```
Same logic, this is very hard to read. Please add comments

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit_schema.hpp:
```diff
@@ -16,10 +16,10 @@ namespace smt_circuit_schema {
  * @param modulus Modulus of the field we are working with
  * @param public_inps Public inputs to the current circuit
- * @param vars_of_interes Map wires indicies to their given names
+ * @param vars_of_interes Map wires indices to their given names
```
interest probably (typo)

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/subcircuits.cpp:
```diff
@@ -2,23 +2,35 @@
+    size_t num_gates = builder.get_num_gates() - start_gate;
+
+    CircuitSchema exported = unpack_from_buffer(builder.export_circuit());
+    return { exported, start_gate, num_gates, { 2 }, { num_gates - 1 } };
```
Pleas add this 2 as a contant or make a comment inside brackets {/\*like_this\*/ 2}. Same in the standard logic circuit

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/subcircuits.test.cpp:
```diff
@@ -0,0 +1,54 @@
+        uint32_t a_gate_idx = range_props.idxs[0];
+        size_t start_gate = range_props.start_gate;
+
+        ASSERT_EQ(
```
Please add a comment what you are doing here

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/subcircuits.test.cpp:
```diff
@@ -0,0 +1,54 @@
+        uint32_t a_gate_idx = logic_props.idxs[0];
+        size_t start_gate = logic_props.start_gate;
+        ASSERT_EQ(
+            "a", circuit.vars_of_interest[circuit.real_variable_index[circuit.wires[start_gate + a_gate][a_gate_idx]]]);
```
Please add a comment what you are doing here

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/smt_examples.test.cpp:
```diff
@@ -153,29 +153,62 @@ TEST(circuit_verifiaction, unique_witness)
-
-TEST(solver_use_case, solver)
+TEST(SMT_Example, unique_witness)
+// two roots of a quadratic eq x^2 + a * x + b = s
```
Please move this line about "TEST" or inside the brackets

Rumata888 replied:
Also, could you please make a more explicit description?

Rumata888 reviewed (changes requested):
Please address my comments

Sarkoxed commented on barretenberg/cpp/src/barretenberg/smt_verification/circuit/circuit.hpp:
```diff
@@ -168,6 +186,168 @@ void Circuit<FF>::univariate_handler(bb::fr q_m, bb::fr q_1, bb::fr q_2, bb::fr
 }
 
+/**
+ * @brief Relaxes univariate logic constraints(AND/XOR).
```
Mistake here

Sarkoxed committed:
Expanded description

Sarkoxed committed:
Typo

Sarkoxed committed:
Elaborate on return values and tests

Sarkoxed committed:
Added test for cache

Sarkoxed committed:
Post review commit

- Changed variable names
- Added cache for subcircuits
- Added more comments
- Renamed handle methods

Sarkoxed committed:
Merge branch 'as/smt-update-circuit-2' of github.com:AztecProtocol/aztec-packages into as/smt-update-circuit-2

Rumata888 committed:
fix small typo

Rumata888 reviewed (approved):
LGTM

Sarkoxed committed:
Merge branch 'master' into as/smt-update-circuit-2

Rumata888 committed:
Merge branch 'master' into as/smt-update-circuit-2
